modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -367,37 +412,54 @@ public CompletableFuture<Void> finish(
             // If the state is FINISHING then someone else hase in in the middle of finishing this tx.
             if (stateMeta.txState() == FINISHING) {
                 return ((TxStateMetaFinishing) stateMeta).txFinishFuture()
-                        .thenCompose(meta -> checkTxOutcome(commit, txId, meta));
+                        .thenCompose(meta -> checkTxOutcome(commitIntent, txId, meta));
             } else {
                 // The TX has already been finished. Check whether it finished with the same outcome.
-                return checkTxOutcome(commit, txId, stateMeta);
+                return checkTxOutcome(commitIntent, txId, stateMeta);
             }
         }
 
-        TxContext tuple = txCtxMap.compute(txId, (uuid, tuple0) -> {
-            if (tuple0 == null) {
-                tuple0 = new TxContext(); // No writes enlisted.
-            }
-
-            assert !tuple0.isTxFinishing() : "Transaction is already finished [id=" + uuid + "].";
-
-            tuple0.finishTx();
-
-            return tuple0;
-        });
+        TxContext txContext = lockTxForNewUpdates(txId, enlistedGroups, commitIntent);
 
         // Wait for commit acks first, then proceed with the finish request.
-        return tuple.performFinish(commit, ignored ->
+        return txContext.performFinish(commitIntent, commit ->
                 prepareFinish(
                         observableTimestampTracker,
                         commitPartition,
+                        commitIntent,
                         commit,
                         enlistedGroups,
                         txId,
                         finishingStateMeta.txFinishFuture()
                 ));
     }
 
+    private TxContext lockTxForNewUpdates(UUID txId, Map<TablePartitionId, Long> enlistedGroups, boolean commitIntent) {
+        TxContext txContext = txCtxMap.compute(txId, (uuid, tuple0) -> {
+            if (tuple0 == null) {
+                tuple0 = new TxContext(); // No writes enlisted.
+            }
+
+            assert !tuple0.isTxFinishing() : "Transaction is already finished [id=" + uuid + "].";
+
+            tuple0.finishTx(enlistedGroups);
+
+            return tuple0;
+        });
+
+        if (commitIntent) {
+            for (Map.Entry<TablePartitionId, Long> e : enlistedGroups.entrySet()) {
+                ReplicaMeta replicaMeta = placementDriver.currentLease(e.getKey());
```

Comment:
This is the point when we still haven't set the commit timestamp, and if we see that some of data nodes' primary replica has expired (which can happen by any reason including node failure), this means that we will not be able to commit this transaction, because `verifyCommitTimestamp` will not pass. But here we need to complete the `waitRepFut` to be able to proceed.

About updating the single field - it's not necessary, we can break the loop after the first exception, I'll fix this